eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
+     elif filter.symKey.isSome():
+       keyHash = keccak256.digest(filter.symKey.get())
+     # else:
+       # NOTE: should we error on messages without encryption?
+     if decoded.isNone():
+       continue
+   else:
+     if filter.privateKey.isSome():
+       if keyHash != keccak256.digest(filter.privateKey.get().data):
+         continue
+     elif filter.symKey.isSome():
+       if keyHash != keccak256.digest(filter.symKey.get()):
+         continue
+     # else:
+       # NOTE: should we error on messages without encryption?
+
+   # When decoding is done we can check the src (signature)
+   if filter.src.isSome():
+     let src: Option[PublicKey] = decoded.get().src
+     if not src.isSome():
+       continue
+     elif src.get() != filter.src.get():
+       continue
+
+   let receivedMsg = ReceivedMessage(decoded: decoded.get(),
+                                     timestamp: msg.env.expiry - msg.env.ttl,
+                                     ttl: msg.env.ttl,
+                                     topic: msg.env.topic,
+                                     pow: msg.pow,
+                                     hash: msg.hash)
+   # Either run callback or add to queue
+   if filter.handler.isSome():
+     filter.handler.get()(receivedMsg)
+   else:
+     filter.queue.insert(receivedMsg)
+
+proc getFilterMessages*(filters: var Filters, filterId: string): seq[ReceivedMessage] =
+  result = @[]
+  if filters.contains(filterId):
+    if filters[filterId].handler.isNone():
+      result = filters[filterId].queue
+      filters[filterId].queue =
+        newSeqOfCap[ReceivedMessage](defaultFilterQueueCapacity)
+
+proc toBloom*(filters: Filters): Bloom =
+  for filter in filters.values:
+    if filter.topics.len > 0:
+      result = result or filter.bloom
+
+type
+  PeerState = ref object
+    initialized*: bool # when successfully completed the handshake
+    powRequirement*: float64
+    bloom*: Bloom
+    isLightNode*: bool
+    trusted*: bool
+    received: HashSet[Message]
+    running*: bool
+
+  WhisperState = ref object
+    queue*: Queue
+    filters*: Filters
+    config*: WhisperConfig
+
+proc run(peer: Peer) {.async.}
+proc run(node: EthereumNode, network: WhisperState) {.async.}
+
+proc initProtocolState*(network: var WhisperState, node: EthereumNode) =
+  network.queue = initQueue(defaultQueueCapacity)
+  network.filters = initTable[string, Filter]()
+  network.config.bloom = fullBloom()
+  network.config.powRequirement = defaultMinPow
+  network.config.isLightNode = false
+  network.config.maxMsgSize = defaultMaxMsgSize
+  asyncCheck node.run(network)
+
+rlpxProtocol shh(version = whisperVersion,
+                 peerState = PeerState,
+                 networkState = WhisperState):
+
+  onPeerConnected do (peer: Peer):
+    debug "onPeerConnected Whisper"
+    let
+      shhNetwork = peer.networkState
+      shhPeer = peer.state
+
+    asyncCheck peer.status(whisperVersion,
```

Comment:
I'll introduce a helper called `exchangeHandshake` that tries to hide the complicated `asyncCheck` and `await` logic below. As we discussed, it seems to be repeated in almost every protocol.

Reply:
Yes, problem can occur on all protocols